### PR TITLE
Stmp sonar

### PR DESCRIPTION
**Pull request by nchengusaf**
Initial commit for first batch of fingerprints

Edit by @tsellers-r7: For anyone reviewing this PR - this PR is part of a learning exercise introducing nchengusaf to Recog, Ruby, and regex.  He's being thrown into the deep end..

### CODE REVIEW

File: xml/smtp_banners.xml [hunk 1]
```diff
@@ -1647,4 +1648,93 @@ The system or service fingerprint with the highest certainty overwrites the othe
     <param pos="0" name="service.product" value="Unknown"/>
     <param pos="1" name="host.name"/>
   </fingerprint>
+
```
Comment (dabdine-r7) [inline xml/smtp_banners.xml]: Get rid of the whitespace

File: xml/smtp_banners.xml [hunk 2]
```diff
@@ -1647,4 +1648,93 @@ The system or service fingerprint with the highest certainty overwrites the othe
...
+
+
+
+
+
+
+
+
+        <fingerprint pattern=”^(\w+.\w+.\w)+ Service not available$>
```
Comment (dabdine-r7) [inline xml/smtp_banners.xml]: looks like unicode quotations are used here, and elsewhere.

Comment (dabdine-r7) [inline xml/smtp_banners.xml]: ” instead of "

File: xml/smtp_banners.xml [hunk 3]
```diff
@@ -1647,4 +1648,93 @@ The system or service fingerprint with the highest certainty overwrites the othe
...
+        <description>match to Service not available error
```
Comment (dabdine-r7) [inline xml/smtp_banners.xml]: all the description elements in this file are missing closing tags (`</description>`)

File: xml/smtp_banners.xml [hunk 4]
```diff
@@ -1647,4 +1648,93 @@ The system or service fingerprint with the highest certainty overwrites the othe
...
+        <example service.version=”4.3.2”> 4.3.2 Service not available <example>
+        <param pos="1" name="service.version"/>
+        <param pos=”0” name=”service.vendor” value=”new”/>
+        <param pos=”0” name=”service.family” value=”version”/>
+        <param pos=”0” name=”service.product” value=”version”/>
+        <fingerprint>
+         <fingerprint pattern="^(.+.{4})-\w.+echosting+[^ ]+(cafe+.+.+.[a-z]+[a-z]+[a-z])+ ESMTP$">
+        <description>match to echosting or ESMTP
+        <example service.version="e0000" website.version="cafe24.com"> e0000-e.echosting.cafe24.com ESMTP <example>
```
Comment (dabdine-r7) [inline xml/smtp_banners.xml]: don't include hostnames from the source sonar data (replace hostnames with something like foo.com, and make sure the regular expression still matches).

File: xml/smtp_banners.xml [hunk 5]
```diff
@@ -1647,4 +1648,93 @@ The system or service fingerprint with the highest certainty overwrites the othe
...
+        <param pos="0" name="service.vendor" value="new"/>
+        <fingerprint>
+         <fingerprint pattern="^([a-z]+[a-z]+[a-z]+.)+.cloud.netart +ESMTP +ready$">
+        <description>match to cloud.netart and ESMTP and ready state
+        <example service.version="lbc4"/> lbc4.cloud.netart ESMTP ready <example>
+        <param pos="1" name="service.version"/>
+        <param pos="0" name="service.vendor" value="new"/>
+        <param pos="0" name="service.family" value="version"/>
+        <fingerprint>
+         <fingerprint pattern="^([A-Z]+[A-Z]+[A-Z]+[A-Z])+[^ ]+(\d+[^ ]+\d) (\d\d\d) +Unauthorized$">
```
Comment (dabdine-r7) [inline xml/smtp_banners.xml]: Remove this fingerprint -- it isn't a mail server, but an HTTP server which was included. All sorts of things run on the internet, sometimes we capture HTTP servers running on mail ports.

File: xml/smtp_banners.xml [hunk 6]
```diff
@@ -1647,4 +1648,93 @@ The system or service fingerprint with the highest certainty overwrites the othe
...
+        <description> match to HTTP error
+        <example service.version="HTTP" service.number="1.0" error.number="403"> HTTP/1.0 403 Unauthorized <example>
```
Comment (dabdine-r7) [inline xml/smtp_banners.xml]: Same comment as above (remove this fingerprint)

File: xml/smtp_banners.xml (hunk 3 above)
Comment (dabdine-r7) [inline xml/smtp_banners.xml]: Also, the fingerprint element here is missing a closing tag

**Comment by tsellers-r7** [comment]
Per comments from Derek, before submitting or updating a PR you should validate that the changes do not cause the tests to fail.

1. `cd <path to clone of recog repo>`
2. Verify that the fingerprints themselves will validate against the examples
    `bin/recog_verify <xml_file_path>`

    For example:
    `bin/recog_verify xml/smtp_banners.xml`

    This may generate many warnings, but no failures.  The warnings are undesirable but don’t break anything. Try not to increase the warning count ;)
3. Run RSpec to run all of the project's unit tests. This will take a while, should complete without errors.

   `rspec`

File: xml/smtp_banners.xml [hunk 7]
```diff
@@ -1647,4 +1648,93 @@ The system or service fingerprint with the highest certainty overwrites the othe
...
+        <param pos="1" name="service.version"/>
+        <param pos="2" name="service.number"/>
+        <param pos="3" name="error.number"/>
+        <fingerprint>
```
Comment (tsellers-r7) [inline xml/smtp_banners.xml]: The above line and similar occurrences at the end of the fingerprint should be a closing `</fingerprint>` tag.

File: xml/smtp_banners.xml (hunk 4 above)
Comment (tsellers-r7) [inline xml/smtp_banners.xml]: I think we've been trying to standardize on `foo.bar` as the hostname.  The hostname  is data that we capture as part of this process but, with a few exceptions, almost no Recog fingerprints will be tailored to the hosting provider.

File: xml/smtp_banners.xml [hunk 8]
```diff
@@ -1647,4 +1648,93 @@ The system or service fingerprint with the highest certainty overwrites the othe
...
+         <fingerprint pattern="^(\D{2}+\d{3}).secure.ne.jp+ ESMTP">
+        <description> match to secure.ne.jp in Japan
+        <example service.version="aa227"/> aa227.secure.ne.jp ESMTP <example>
+        <param pos="1" name="service.version" />
+        <param pos="0" name="service.vendor" value="japan" />
+        <param pos="0" name="service.family" value="secure"/>
+        <fingerprint>
+         <fingerprint pattern="^(\D+\D+\d+\d).([a-z]+[a-z]+[a-z]+[a-z]+[a-z])+.jp+ ESMTP$">
+        <description> match to domain controller
+        <example service.version="dc51"/> dc51.etius.jp ESMTP <example>
```
Comment (tsellers-r7) [inline xml/smtp_banners.xml]: Be careful about the assertions and assumptions that are expressed as part of a fingerprint.  For example, are you sure that this is a domain controller and that `dc` doesn't stand for Data Center instead?  The `service.vendor` and `service.family` values below are incorrect as well unless there is a vendor named `Japan` with a product family named `etius`.

**Pull request closed without merging**